MAVProxy/modules/mavproxy_ais.py

```diff
@@ -0,0 +1,150 @@
+            threat_radius = 63
+        else:
+            threat_radius = max(self.state.dimension_bow, self.state.dimension_stern, self.state.dimension_port, self.state.dimension_starboard)
+        #default to threat_radius setting
+        if threat_radius == 0:
+            return default_radius
+
+class AISModule(mp_module.MPModule):
+
+    def __init__(self, mpstate):
+        super(AISModule, self).__init__(mpstate, "ais", "AIS data support", public=True)
+        self.threat_vehicles = {}
+
+        self.add_command('ais', self.cmd_AIS, "ais control",
+                         ["<status>", "set (AISSETTING)"])
+
+        self.AIS_settings = mp_settings.MPSettings([("timeout", int, 60),  # seconds
+                                                    ("threat_radius", int, 200)])  # meters
+        self.add_completion_function('(AISSETTING)',
+                                     self.AIS_settings.completion)
+
+        self.threat_timeout_timer = mavutil.periodic_event(2)
+
+        self.update_map_timer = mavutil.periodic_event(1)
+
+        self.tnow = self.get_time()
+
+    def cmd_AIS(self, args):
+        '''ais command parser'''
+        usage = "usage: ais <set>"
+        if len(args) == 0:
+            print(usage)
+        elif args[0] == "set":
+            self.AIS_settings.command(args[1:])
+        else:
+            print(usage)
+
+    def check_threat_timeout(self):
+        '''check and handle threat time out'''
+        for id in self.threat_vehicles:
+            if self.threat_vehicles[id].update_time == 0:
+                self.threat_vehicles[id].update_time = self.get_time()
+            dt = self.get_time() - self.threat_vehicles[id].update_time
+            if dt > self.AIS_settings.timeout:
+                # if the threat has timed out...
+                del self.threat_vehicles[id]  # remove the threat from the dict
+                for mp in self.module_matching('map*'):
+                    # remove the threat from the map
+                    mp.map.remove_object(id)
+                    mp.map.remove_object(id+":circle")
+                # we've modified the dict we're iterating over, so
+                # we'll get any more timed-out threats next time we're
+                # called:
+                return
+
+    def update_map(self):
+        '''update the map graphics'''
+        for mp in self.module_matching('map*'):
+            for id in self.threat_vehicles.keys():
+                mstate = self.threat_vehicles[id].state
+                threat_radius = self.threat_vehicles[id].getThreatRadius(self.AIS_settings.threat_radius)
+                # update if existing object on map, else create
+                if self.threat_vehicles[id].onMap:
+                    mp.map.set_position(id, (mstate.lat * 1e-7, mstate.lon * 1e-7), 3, rotation=mstate.heading*0.01, label=str(mstate.callsign))
+                    # Turns out we can't edit the circle's radius, so have to remove and re-add
```

Comment:
How expensive is this?  Should we check the threat radius has changed before doing this?